draft-ietf-mls-protocol.md

```diff
@@ -1452,11 +1452,43 @@ message with invalid values for public keys in the ratchet tree. ]]
 
 ## Init
 
-[[ OPEN ISSUE: Direct initialization is currently undefined.  A client can
-create a group by initializing its own state to reflect a group
-including only itself, then adding the initial members.  This
-has computation and communication complexity O(N log N) instead of
-the O(N) complexity of direct initialization. ]]
+A group can always be created by initializing a one-member group and
+using adding members individually.  For cases where the initial list
+of members is known, the Init message allows a group to be created
+more efficiently.
+
+~~~~~
+struct {
+  opaque group_id<0..255>;
+  ProtocolVersion version;
+  CipherSuite cipher_suite;
+  UserInitKey members<0..2^32-1>;
+  DirectPath path;
```

Comment:
You would actually want a giant vector of `DirectPathNode` values, where each private key was encrypted to the `2^level` receipients below it.  But I agree we should punt this to a future version.

Reply:
I think the current scenario looks fine. In the scenario were we would break the secrecy invariant and allow warmup and pre-populating nodes, there is an performance optimization: we can just send the public and the secret part of the tree independently to avoid large numbers of PKE of a large payload. In that scenario I would suggest encrypting the pre-warmed public tree encrypted under a fresh AEAD key/nonce and PKE encrypt the key and the nonce to the recipient along with the new path secret of its parent.